runsc/container/container.go

```diff
@@ -287,65 +290,82 @@ func New(conf *config.Config, args Args) (*Container, error) {
 		if err != nil {
 			return nil, fmt.Errorf("error creating pod mount hints: %w", err)
 		}
-		goferFilestores, goferConfs, err := c.createGoferFilestores(conf.GetOverlay2(), mountHints)
+		rootfsHint, err := boot.NewRootfsHint(args.Spec)
+		if err != nil {
+			return nil, fmt.Errorf("error creating rootfs hint: %w", err)
+		}
+		goferFilestores, goferConfs, err := c.createGoferFilestores(conf.GetOverlay2(), mountHints, rootfsHint)
 		if err != nil {
 			return nil, err
 		}
+		GPUFunctionalityRequested := specutils.GPUFunctionalityRequested(args.Spec, conf)
+		if GPUFunctionalityRequested && !goferConfs[0].ShouldUseLisafs() {
+			// TODO: Support GPU functionality without LISAFS backed root mount.
+			return nil, fmt.Errorf("GPU functionality is not available when using non-LISAFS backed root mount")
+		}
```

Comment:
Hmm why is it not supported? We can spawn the gofer process with no `io-fds` right? I think we just need a method like:
```
func shouldSpawnGofer(spec, conf) bool {
    if rootfs is lisafs {
        return true
    }
    if any gofer mount present in spec.Mounts as determined by specutils.IsGoferMount() {
        return true
    }
    if specutils.GPUFunctionalityRequested(spec, conf) {
        return true
    }
    return false
}
```

Maybe call this from `Container.createGoferProcess()` and return no IO files if gofer is not needed.

Reply:
Makes sense. If we only need to proxy the char devices, we just need to spawn the gofer process. But `nvidia-container-cli` will bind mount host files (e.g. `nvidia-smi`) into container's rootfs. And when EROFS rootfs is used, these files won't be available inside the container.. Maybe I just need to return an error only when `--nvproxy-docker` is requested. I will do that in the next version.

Reply:
`nvidia-container-cli` will bind mount the GPU devices into `spec.Root.Path` directory. But that directory is never exposed to the application via the rootfs. We have a mandatory mount of type `fsimpl/dev` at `/dev`.
https://github.com/google/gvisor/blob/aeaee716690083700b7d500c7bec1d36a161ba1a/runsc/boot/vfs.go#L263-L268

So it is okay for EROFS to not show these devices. GPU devices are exposed to the sandbox by bind mounting them into the sandbox process:
https://github.com/google/gvisor/blob/aeaee716690083700b7d500c7bec1d36a161ba1a/runsc/cmd/chroot.go#L186-L206

And then the sandbox directly opens these devices. I have work in the pipeline where we will not do these sandbox bind mounts, and instead open a gofer connection which will serve these devices from `spec.Root.Path` directory. See https://github.com/google/gvisor/pull/9624.

Reply:
Oh, you are talking about bind mounting host files (not host GPU devices). Hmm yeah that's an issue. Making EROFS incompatible with `--nvproxy-docker` SGTM for now.